src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -202,9 +284,173 @@ public Task LoadAsync()
 
         public Task UnloadAsync()
         {
-            _enabled = false;
+            lock (SyncObject)
+            {
+                _enabled = false;
+
+                _whenNominatedTask?.TrySetCanceled();
+            }
 
             return Task.CompletedTask;
         }
+
+        // NuGet calls this method to wait project to nominate restoring.
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project nominate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to nominate a restore (DT build failed)
+        public Task WhenNominated(CancellationToken cancellationToken)
+        {
+            lock (SyncObject)
+            {
+                if (!CheckIfHasPendingNomination() || cancellationToken.IsCancellationRequested)
+                {
+                    return Task.CompletedTask;
+                }
+
+                _whenNominatedTask ??= new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+                if (!_wasSourceBlockContinuationSet)
+                {
+                    _wasSourceBlockContinuationSet = true;
+
+                    _ = SourceBlock.Completion.ContinueWith(t =>
+                    {
+                        lock (SyncObject)
+                        {
+                            if (_whenNominatedTask is not null)
+                            {
+                                if (t.IsFaulted)
+                                {
+                                    _whenNominatedTask.SetException(t.Exception);
+                                }
+                                else
+                                {
+                                    _whenNominatedTask.TrySetCanceled();
+                                }
+                            }
+                        }
+                    }, CancellationToken.None, TaskContinuationOptions.OnlyOnFaulted, TaskScheduler.Default);
+                }
+            }
+
+            return _whenNominatedTask.Task.WithCancellation(cancellationToken);
+        }
+
+        // True means the project system plans to call NominateProjectAsync in the future.
+        bool IVsProjectRestoreInfoSource.HasPendingNomination
```

Comment:
Merge HasPendingNomination into CheckIfHasPendingNomination, it's an unneeded concept indirection.

Reply:
fixed in this [commit ](https://github.com/dotnet/project-system/pull/7199/commits/d5a2a1f9b385e6a96d6d1c5fcbde5b7d0f19ebc5)